### ISSUE TITLE
support HAR exporter (http) layer in rama

### ISSUE THREAD

**Issue by GlenDC** [plabayo/rama#357]
By default it exports to a file of your choice,
but we can also support to provide custom writers I suppose :)

Most likely we need to do the actual writing over a channel as to ensure there's only ever one active writer to the file though, given a (rama) service is `Send+Sync`.

HAR File wikipedia: https://en.wikipedia.org/wiki/HAR_(file_format)

**Comment by hafihaf123** [comment]
Hello,

I am very interested in contributing to this issue. Although I am new to Rust and this would be my first time contributing to open-source, I am eager to learn and would greatly appreciate any guidance to help me address this task. While I will do my best to make progress promptly, I may require additional time and support to ensure the solution meets the project’s standards.

Thank you for considering my request!

**Comment by GlenDC** [comment]
Hi @hafihaf123 this issue is still available, it might however be a bit much both in terms of complexity and the amount of work required if I read your self described background.

I just created a new ticket. Would you be interested in try that one (first)? It also has me as a mentor available.

https://github.com/plabayo/rama/issues/358

If you want to pick that one up you can let it be known in that issue.

Nice to meet you btw!

**Comment by Hrushi20** [comment]
Hey! I'm interested in working on this issue. Looks fun to solve it.

**Comment by GlenDC** [comment]
Yes, but I would prefer that you finish your existing issue prior to starting a new one.
If by then it is still available it is all yours.

**Comment by GlenDC** [comment]
Do ask if something is not clear btw @Hrushi20 or if you need help/guidance in any way. This is by no means a small issue :) But at the same time be aware of scope creep. The goal of this issue is minimal enough.

**Comment by ASamedWalker** [comment]
Hi,
I am very interested in contributing to this issue. Although I am new to Rust and this would be my first time contributing to open-source, I am eager to learn and would greatly appreciate any guidance to help me address this task. While I will do my best to make progress promptly, I may require additional time and support to ensure the solution meets the project’s standards. Thank you for considering my request!

**Comment by GlenDC** [comment]
I believe @Hrushi20 was going to pick this one up already. Is that still the case @Hrushi20 ?

**Comment by Hrushi20** [comment]
Yeah. I was going through the rama documentation to understand things on a high level

**Comment by Hrushi20** [comment]
Hey! I just wanted to clarify my understanding about problem statement: 

* The trigger trait which needs to be created takes a function as input. This function is responsible for starting/stopping HAR file recording. We control the Trigger Trait via the Layer. The Layer communicates the Trigger to the Service. 
* The HAR service is spawned as a new task listening to requests. 
* The HAR service internally has an Atomic Boolean. Toggling the boolean ensures tracking the state of recording of HAR files. What does the stopped modulus mean here? 
* The HAR service is also responsible for observing incoming Requests, validate if it's HAR file. (Or can we assume HAR service always receives HAR files as part of request?) and write the data using a writer. 
* Will create a [HAR data structs](https://w3c.github.io/web-performance/specs/HAR/Overview.html) as per the w3c spec.

**Comment by GlenDC** [comment]
Hmmm you need to take a step back I think. 

Take a look at <https://github.com/plabayo/rama/blob/main/rama-http/src/layer/traffic_writer/request.rs>, as the `HarService` and `HarLayer` are very similar to this.

the `HarLayer` is a `rama::Layer` which produces a `HarService` (struct) that implements `rama::Service`. That's the basic gist of it. And the `HarService` `Service` implementation would be in pseudo-real code something like this:

```rust
impl<State, S, W, ReqBody, ResBody> Service<State, Request<ReqBody>> for HarService<S, W>
where
    State: Clone + Send + Sync + 'static,
    S: Service<State, Request, Response = Response<ResBody>, Error: Into<BoxError>>,
    W: RequestWriter,
    ReqBody: http_body::Body<Data = Bytes, Error: Into<BoxError>> + Send + Sync + 'static,
    ResBody: Send + 'static,
{
    type Response = Response<ResBody>;
    type Error = BoxError;

    async fn serve(
        &self,
        ctx: Context<State>,
        req: Request<ReqBody>,
    ) -> Result<Self::Response, Self::Error> {
        match self.recorder() {
             Some(recorder) => {
                     recorder.record_request(&req)?;
                     let response = self.inner.serve(ctx, req).await?;
                     recorder.record_response(&response)?;
                     Ok(response)
             },
             None => Ok(self.inner.serve(ctx, req).await?),
        }
    }
}
```

Just an example though, but the high level idea is as simple as that.

The recorder would be just something internally nothing we expose as this is not a generic service. Whether or not there is an active recorder depends whether or not we are recording.

So the layer and service could have the following configuration options:

- Toggle: this would be a simple trait and that we can also `SignalToggle`, more about this later
- HarSink:  a trait, more about this later
- Guard: a optional <https://ramaproxy.org/docs/rama/graceful/struct.ShutdownGuard.html> in case you want this to be gracefully spawned, otherwise you would use `tokio::spawn`

And that's about it.

the `Recorder` would be there to collect the request and send it together with the response to the internal collector when it drops. This also mean it can send it as just the request in case there was never a response, that's fine as well. But it's something internal to the har service no need to expose this.

When `HarService` gets created you would spawn a task (using the `Guard` or `tokio::spawn`) which has a loop:

```
loop {
     let mut active = false;
      tokio::select! {
          _ = shutdown => { /* clean up + exit */ }
          _ = toggle.toggle() => { active = !active }
      }

      if active { /* ... */ } else { /* .... */ }
}
```

(it's in this loop that you'll use the configured `HarSink` to send the `har` object to upon exit or toggle-off).

`Toggle` could be as simple as:

```
trait Toggle {
       fn toggle(&self) => impl Future + Send + '_; // needs to be cancel safe
}
```

There's for example Signal in tokio: https://docs.rs/tokio/latest/tokio/signal/unix/struct.Signal.html

so we can easily implement already Toggle for it as:

```rust
#[if on unix]
impl Toggle for tokio::signal::unix::Signal {
      #[inline]
      async fn toggle(&self)  {
           self.recv().await
      }
}
```

and same for the windows ones if on windows

You can also implement it already for https://docs.rs/tokio/latest/tokio/sync/mpsc/index.html's Receiver

and finally you can already impleemnt it for any `Fn` as well which returns a future.

Finally the `HarSink` would be something like:

```rust
trait HarSink  {
      // only send error if it's a fatal error, this will make sure the loop exits and the `HarService` will
      // from this point onward never be in record-modus anymore
      fn send_har(&mut self, har: Har) -> impl Future<Output = Result<(), Self::Error>> 
}
```

You can already implement this `HarSink` for:

- `mpsc::Sender<Har>`: just send it
- `tokio::io::AsyncWrite` serialize the `Har` object as bytes and write it in the writer

And that's about it.

I hope this makes it a bit more clear.

I would say keep it simple for now though as there's a lot of information that can written to a har file that you don't have information yet on and that do not need to be in this pass. Some extra things that you can already do are:

- provide a `HarComment` struct that can be created as `HarComment::new(fmt::Display)` which can be injected into a `Context<S>'s Extensions` so you can add it as a `"comment"` property for that request/response pair's entry object :)

But some things are just a lot harder for now and wouldn't worry about it:

- e.g. there are the timing objects which can have a lot of granularity (including dns, etc... I would justl eave a lot or most of that empty
- etc etc

See this ticket mostly as an MVP (Minimal Viable Product). We can always expand the support feature set of Har as time progresses and the need is there for it. For now if the basic timings and the the recording of the request and response objects are already there it would already be a pretty great start.

**Comment by Hrushi20** [comment]
I've started implementation, I was thinking about Sink, Toggle Trait. As per my understanding, 

* Toggle Trait is used for Clients to Enable/Disable Har Service. Toggle Trait Ensure the Recorder is set to enable recording in the Service. 
* Sink Trait is used to finally output the data when toggle (false)/ shutting down the process.

**Comment by GlenDC** [comment]
Yes, @Hrushi20, that sounds about right. Think of it like a record button on a tape recorder. Of course, nothing stops you from running it indefinitely, but that's a bit futile since it will all be stored in volatile memory.

Browser dev tools work the same way; at least Chrome allows you to click "record" and "stop recording," which makes sense to me.

Sadly, HAR is a bit of a silly format because it can't really be streamed, as it's all one large object. Of course, you could stream it by working with an intermediate export step, so there are possibilities, but for this initial implementation, it's fine to just keep it in RAM.

If there's ever a need for it to work in long-running sessions via streaming, we can address that issue at the appropriate time later on.

**Comment by GlenDC** [comment]
@Hrushi20 do you still want to handle this or do you prefer to hand this over? I don't mind either way, but I do need issues like these to move forward. Whatever the case let me know, and take care. Best wishes.

**Comment by MeerKatDev** [comment]
Hi @GlenDC was just looking for some first issue (saw you on This Week in Rust!), can I be assigned this?

**Comment by GlenDC** [comment]
All yours @MeerKatDev
